ReactiveCocoa/NSObject+Lifetime.swift

```diff
@@ -15,10 +15,53 @@ extension Reactive where Base: NSObject {
 			let token = Lifetime.Token()
 			let lifetime = Lifetime(token)
 
+			let objcClass: AnyClass = (base as AnyObject).objcClass
+			let deallocSelector = sel_registerName("dealloc")!
+
+			// Swizzle `-dealloc` so that the lifetime token is released at the
+			// beginning of the deallocation chain, and only after the KVO `-dealloc`.
+			objc_sync_enter(objcClass)
+			if objc_getAssociatedObject(objcClass, &lifetimeKey) == nil {
+				objc_setAssociatedObject(objcClass, &lifetimeKey, true, .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
+
+				var existingImpl: IMP? = nil
+
+				let newImplBlock: @convention(block) (UnsafeRawPointer) -> Void = { objectRef in
+					_rac_objc_setAssociatedObject(objectRef, &lifetimeTokenKey, nil, .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
+
+					let impl: IMP
+
+					if let existingImpl = existingImpl {
+						impl = existingImpl
+					} else {
+						let superclass: AnyClass = class_getSuperclass(objcClass)
+						impl = class_getMethodImplementation(superclass, deallocSelector)
+					}
+
+					typealias Impl = @convention(c) (UnsafeRawPointer, Selector) -> Void
+					unsafeBitCast(impl, to: Impl.self)(objectRef, deallocSelector)
+				}
+
+				let newImpl =  imp_implementationWithBlock(newImplBlock as Any)
+
+				if !class_addMethod(objcClass, deallocSelector, newImpl, "v@:") {
+					// The class has an existing `dealloc`. Preserve that as `existingImpl`.
+					let deallocMethod = class_getInstanceMethod(objcClass, deallocSelector)
+					existingImpl = method_getImplementation(deallocMethod)
+					existingImpl = method_setImplementation(deallocMethod, newImpl)
```

Comment:
It is intentional, because while the call itself is thread safe, these calls could in theory race with invocations and other swizzling frameworks. So setting it the first time is to ensure there is at least an implementation, and setting it the second time is to ensure it is up-to-date.

Maybe I should have brought over the comments too.